src/dep.ml

```diff
@@ -4,47 +4,68 @@ module T = struct
   type t =
     | Env of Env.Var.t
     | File of Path.t
+    | Glob of Path.t * Path.t Predicate.t
     | Universe
 
   let env e = Env e
   let file f = File f
   let universe = Universe
+  let glob ~dir predicate = Glob (dir, predicate)
 
   let compare x y =
     match x, y with
     | Env x, Env y -> Env.Var.compare x y
     | Env _, File _
+    | Env _, Glob _
     | Env _, Universe -> Ordering.Lt
     | File x, File y -> Path.compare x y
+    | File _, Glob _
     | File _, Env _ -> Ordering.Gt
     | File _, Universe -> Ordering.Lt
+    | Glob (d1, p1), Glob (d2, p2) ->
+      Tuple.T2.compare Path.compare Predicate.compare (d1, p1) (d2, p2)
+    | Glob _, (Env _ | File _) -> Ordering.Gt
```

Comment:
(File, Glob) gives Gt, and yet (Glob, File) also gives Gt?
If we're writing it manually, maybe we should write it in a less error-prone way? Something like this perhaps?

```
| Env _, Env _ -> ...logic...
| Env _, _ -> Lt
| _, Env -> Gt

| File _, File _ -> ...logic...
| File _, _ -> Lt
| _, File -> Gt
```

... etc

Reply:
Good catch. Thanks. I'm getting sleepy writing all of these equality/comparison/hashing/sexp functions.